src/appleseed/renderer/modeling/bsdf/glossybrdf.cpp

```diff
@@ -307,24 +407,99 @@ namespace
                 alpha_x,
                 alpha_y);
 
-            const float gamma =
-                highlight_falloff_to_gama(values->m_highlight_falloff);
-
-            return
-                MicrofacetBRDFHelper::pdf(
-                    *m_mdf,
-                    alpha_x,
-                    alpha_y,
-                    gamma,
-                    shading_basis,
-                    flipped_outgoing,
-                    flipped_incoming);
+            switch (m_mdf_type)
+            {
+                case GGX:
+                {
+                    const GGXMDF mdf;
+                    return MicrofacetBRDFHelper::pdf(
+                        mdf,
+                        alpha_x,
+                        alpha_y,
+                        1.0f,
+                        shading_basis,
+                        flipped_outgoing,
+                        flipped_incoming);
+                }
+                break;
+
+                case Beckmann:
+                {
+                    const BeckmannMDF mdf;
+                    return MicrofacetBRDFHelper::pdf(
+                        mdf,
+                        alpha_x,
+                        alpha_y,
+                        1.0f,
+                        shading_basis,
+                        flipped_outgoing,
+                        flipped_incoming);
+                }
+                break;
+
+                case Std:
+                {
+                    const StdMDF mdf;
+                    return MicrofacetBRDFHelper::pdf(
+                        mdf,
+                        alpha_x,
+                        alpha_y,
+                        highlight_falloff_to_gama(values->m_highlight_falloff),
+                        shading_basis,
+                        flipped_outgoing,
+                        flipped_incoming);
+                }
+                break;
+
+                assert_otherwise;
+            }
         }
 
       private:
         typedef GlossyBRDFInputValues InputValues;
 
-        unique_ptr<MDF> m_mdf;
+        template <typename MDF>
+        static void add_energy_compensation_term(
+            const MDF&                  mdf,
+            const InputValues*          values,
+            const float                 cos_in,
+            const float                 cos_on,
+            Spectrum&                   value)
+        {
+            if (values->m_energy_compensation != 0.0f)
+            {
+                float fms;
+                float eavg;
+                microfacet_energy_compensation_term(
+                    mdf,
+                    values->m_roughness,
+                    cos_in,
+                    cos_on,
+                    fms,
```

Comment:
What do `fms` and `eavg` mean?

Reply:
It's terminology from the paper. fms is the multiscattering part of the BSDF (the energy we want to recover). Eavg is the average amount of energy reflected for a specific roughness value.

Reply:
Cool, thanks.